File: src/test/java/kitchenpos/menus/tobe/fixtures/FakeMenuRepository.java
package kitchenpos.menus.tobe.fixtures;

import kitchenpos.menus.tobe.domain.Menu;
import kitchenpos.menus.tobe.domain.MenuRepository;

import java.util.*;

public class FakeMenuRepository implements MenuRepository {

    private Map<UUID, Menu> inMemory = new HashMap<>();

    @Override
    public Menu save(Menu menu) {
        inMemory.put(menu.getId(), menu);
        return inMemory.get(menu.getId());
    }

    @Override
    public Optional<Menu> findById(UUID id) {
        return Optional.ofNullable(inMemory.get(id));
    }

    @Override
    public List<Menu> findAll() {
        return inMemory.values().stream().toList();
    }

    @Override
    public List<Menu> findAllByIdIn(List<UUID> ids) {
        return inMemory.values().stream().filter(menu -> ids.contains(menu.getId())).toList();
    }

    @Override
    public List<Menu> findAllByProductId(UUID productId) {
        return inMemory.values()
                .stream()
                .filter(menu ->
                        menu.getMenuProducts()
                            .stream()
                            .anyMatch(menuProduct -> menuProduct.getProductId().equals(productId)))
                .toList();
    }
}

File: src/test/java/kitchenpos/menus/application/tobe/FakeMenuRepository.kt
package kitchenpos.menus.application.tobe

import kitchenpos.menus.tobe.domain.Menu
import kitchenpos.menus.tobe.domain.MenuRepository
import java.util.*

class FakeMenuRepository : MenuRepository {
    private val menuMap: MutableMap<UUID, Menu> = mutableMapOf()

    override fun save(menu: Menu): Menu {
        menuMap[menu.id] = menu

        return menu
    }

    override fun findAllByProductId(productId: UUID): List<Menu> {
        return menuMap.filterValues { it.menuProducts.productIds.contains(productId) }
            .values.toList()
    }

    override fun findById(menuId: UUID): Optional<Menu> {
        return Optional.ofNullable(menuMap[menuId])
    }

    override fun findByIdIn(menuIds: List<UUID>): List<Menu> {
        return menuIds.mapNotNull { menuMap[it] }
    }
}
